### PR TITLE
Source Facebook Marketing (no-singer)

### PR DESCRIPTION
**Pull request by eugene-kulak**
## Contributor Checklist
Thanks for contributing to Airbyte! Please complete the following items in order so we can review your PR.
- [x] Followed all the instructions in the locally generated checklist and your connector is functional & ready for review
- [x] Ran the standard test suite locally via `./gradlew :airbyte-integrations:connectors:source-<your_source_name>:standardSourceTestPython` and pasted the summarized output as a comment in this PR

## Reviewer Pre-merge Checklist 
- [x] Finished iterating with the PR author on the code*
- [x] Created a branch off master to merge this PR into*
- [x] Inject the credentials in CI via `./tools/integrations/ci_credentials.sh` and `.github/workflows/test-command.yml`*
- [x] Added the credentials for this integration to Github secrets 
- [x] Run standard tests on this branch by commenting `/test connector=<name>`*
- [x] Add entry in `airbyte-config/init/src/main/resources/seed/source_definitions.yaml` to use the new source in Airbyte core
- [x] Deployed the connector to Dockerhub via `./tools/integrations/manage.sh publish airbyte-integrations/connectors/source-<name>`

## Documentation
- [x] Add docs in `docs/integrations/sources/` folder in line with the documentation template found in `docs/contributing-to-airbyte/templates/integration-documentation-template.md`.
- [x] Add link to create docs file to `docs/SUMMARY.md`
- [x] Include a link to the documentation in the `README.md`

### CODE REVIEW

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/client.py [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+
+    # FIXME: filter fields
+    @staticmethod
+    def _get_fields_from_stream(stream: AirbyteStream) -> List[str]:
+        print(stream.json_schema)
+        return []
+
+    def read_stream(self, stream: AirbyteStream) -> Generator[AirbyteRecordMessage, None, None]:
+        """Yield records from stream"""
+        method = self._stream_methods.get(stream.name)
```
Comment (arhip11) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/client.py:304]: I would consider moving this to the base class.
`method = self.get_method(stream.name)` would be more concise and expressive.

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/client.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+    def health_check(self) -> Tuple[bool, str]:
+        return True, "Error"
+
+    # FIXME: filter fields
+    @staticmethod
+    def _get_fields_from_stream(stream: AirbyteStream) -> List[str]:
+        print(stream.json_schema)
+        return []
+
+    def read_stream(self, stream: AirbyteStream) -> Generator[AirbyteRecordMessage, None, None]:
```
Comment (arhip11) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/client.py:302]: Why do you override base class method? Looks like the implementation is identical.

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/client.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+        yield from AdCreativeAPI(self).list(fields)
+
+    def stream__ads_insight(self, fields):
+        client = AdsInsightAPI(self, start_date=self._start_date)
+        yield from client.list(fields)
+
+    def stream__ads_insight_age_and_gender(self, fields):
+        breakdowns = ['age', 'gender']
+        client = AdsInsightAPI(self, start_date=self._start_date, breakdowns=breakdowns)
+        yield from client.list(fields)
```
Comment (arhip11) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/client.py:261]: What if implement this method (and the below ones) with forwarding call to `stream__ads_insight`:
```
yield from self.stream__ads_insight(fields, breakdowns=['age', 'gender'])
```
`stream__ads_insights` signature to be modified though by adding `**kwargs`

**Comment by eugene-kulak** [comment]
```
Test run finished after 36085 ms
[         2 containers found      ]
[         0 containers skipped    ]
[         2 containers started    ]
[         0 containers aborted    ]
[         2 containers successful ]
[         0 containers failed     ]
[         7 tests found           ]
[         0 tests skipped         ]
[         7 tests started         ]
[         0 tests aborted         ]
[         7 tests successful      ]
[         0 tests failed          ]
```

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/common.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+def retry_pattern(backoff_type, exception, **wait_gen_kwargs):
+    def log_retry_attempt(details):
+        _, exc, _ = sys.exc_info()
+        logger.info(exc)
+        logger.info('Caught retryable error after %s tries. Waiting %s more seconds then retrying...',
+                    details["tries"],
+                    details["wait"])
+
+    def should_retry_api_error(exception):
```
[the remark this answers is not in the document]
Comment (eugene-kulak) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/common.py]: I can't find any good reason to reuse these callbacks

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/common.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        if isinstance(exception, FacebookRequestError):
+            return exception.api_transient_error() or exception.api_error_subcode() == 99
```
[the remark this answers is not in the document]
Comment (eugene-kulak) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/common.py]: added constant, unfortunately 99 code is undocumented in FB API

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/client.py [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+
+    def stream__ads_insight_dma(self, fields):
+        breakdowns = ['dma']
+        client = AdsInsightAPI(self, start_date=self._start_date, breakdowns=breakdowns)
+        yield from client.list(fields)
+
+    @staticmethod
+    def _find_account(account_id: str):
+        accounts = fb_user.User(fbid='me').get_ad_accounts()
```
[the remark this answers is not in the document]
Comment (eugene-kulak) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/client.py]: could give some example?
I think the current version is not too difficult to read, but maybe I'm wrong...

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/client.py (hunk 1 above)
Comment (eugene-kulak) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/client.py]: moved to base class, and fixed existing derived classes to accept `kwargs`

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/client.py (hunk 2 above)
Comment (eugene-kulak) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/client.py]: to pass fields, please see my updates

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/client.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (arhip11) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/client.py]: The current code is OK to read, agree. I was thinking about:
`return next((x for x in accounts if x["account_id"] == account_id),)`

[the remark this answers is not in the document]
Comment (eugene-kulak) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/client.py]: well, I like this magic, but not in the code :)

File: docs/integrations/sources/facebook-marketing.md [hunk 7]
```diff
@@ -0,0 +1,85 @@
+# Facebook Marketing
+
+## Sync overview
+
+This source can sync data for the core Ad Campaign data available[ in the Facebook Marketing API](https://developers.facebook.com/docs/marketing-api/campaign-structure): Campaigns, AdSets, Ads, and AdCreatives.
```
Comment (jrhizor) [inline docs/integrations/sources/facebook-marketing.md:5]:
```suggestion
This source can sync data for the core Ad Campaign data available in the [Facebook Marketing API](https://developers.facebook.com/docs/marketing-api/campaign-structure): Campaigns, AdSets, Ads, and AdCreatives.
```

File: airbyte-integrations/connectors/source-freshdesk/.dockerignore [hunk 8]
```diff
@@ -0,0 +1,7 @@
+*
```
Comment (jrhizor) [inline airbyte-integrations/connectors/source-freshdesk/.dockerignore:1]: Should this file be in this PR?

File: airbyte-integrations/connectors/source-facebook-marketing/unit_tests/test_client.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+import pytest
+from airbyte_protocol import AirbyteStream
+from source_facebook_marketing.client import Client, FacebookAPIException
+
+
+def test__heal_check_with_wrong_token():
```
Comment (jrhizor) [inline airbyte-integrations/connectors/source-facebook-marketing/unit_tests/test_client.py:30]:
```suggestion
def test__health_check_with_wrong_token():
```

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/client.py [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+                rec = obj.export_all_data()
+                yield rec
+
+    def _params(self, fields: Sequence[str] = None) -> Iterator[dict]:
+        buffered_start_date = self.start_date.subtract(days=self.buffer_days)
+        end_date = datetime.now()
+
+        fields = list(set(fields) - set(self.INVALID_INSIGHT_FIELDS))
+
+        # Some automatic fields (primary-keys) cannot be used as 'fields' query params.
```
Comment (jrhizor) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/client/client.py:216]: This comment should be moved to where `INVALID_INSIGHT_FIELDS` is defined.

**Review by jrhizor — COMMENTED** [review]
This generally looks good. Just a few minor comments for now. 

I'm going to run some tests on this later tonight or tomorrow morning.

File: airbyte-integrations/connectors/source-freshdesk/.dockerignore (hunk 8 above)
Comment (eugene-kulak) [reply airbyte-integrations/connectors/source-freshdesk/.dockerignore]: deleted

**Review by jrhizor — CHANGES_REQUESTED** [review]
The integration tests are working and everything looks good, but there are errors when I run the unit tests with `./gradlew :airbyte-integrations:connectors:source-facebook-marketing:unitTest` from the airbyte root directory or `.venv/bin/python -m pytest unit_tests` from the source's directory. 

Messages must be strings for `AirbyteLogMessage`.

This is the only issue blocking this PR.

**Comment by eugene-kulak** [comment]
@jrhizor thanks, I have fixed logger usage in the last commit

**Comment by jrhizor** [comment]
You'll also have to run `./gradlew format` to format the JSON files added in the PR.

**Comment by jrhizor** [comment]
I'm running the build again from https://github.com/airbytehq/airbyte/pull/1525 

Please look into build failures if there are any. If it succeeds I'll merge this.

Sorry for the delay.

**Comment by jrhizor** [comment]
It looks like the changes here introduced a dependency issue as well. 

It's really slow for me to create a separate PR and wait for builds to iterate with feedback. Can you reopen this PR on an origin branch instead of a fork so the builds can run automatically? Then you can iterate until the build is passing without any involvement from me, and it will be obvious when this is ready to merge.

**Comment by eugene-kulak** [comment]
closed in favor of #1552

**Pull request closed without merging**